Meta registrations, Nov 7 edition, part 1

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #88595
* __->__ #88584

I already had to fix a bunch of bugs in this batch, so landing it for
now before I tackle the next batch.

Billing of changes:
- diag.out is now composite
- randint layout/pin_memory bugfix
- unsqueeze_ meta
- bernoulli metas
- mul_/div_/logical_{and_,or_,xor_} metas
- add_/sub_ metas with alpha support
- diag_embed.out meta

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Meta registrations, Nov 7 edition, part 1

I already had to fix a bunch of bugs in this batch, so landing it for
now before I tackle the next batch.

Billing of changes:
- diag.out is now composite
- randint layout/pin_memory bugfix
- unsqueeze_ meta
- bernoulli metas
- mul_/div_/logical_{and_,or_,xor_} metas
- add_/sub_ metas with alpha support
- diag_embed.out meta

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Meta registrations, Nov 7 edition, part 1"


I already had to fix a bunch of bugs in this batch, so landing it for
now before I tackle the next batch.

Billing of changes:
- diag.out is now composite
- randint layout/pin_memory bugfix
- unsqueeze_ meta
- bernoulli metas
- mul_/div_/logical_{and_,or_,xor_} metas
- add_/sub_ metas with alpha support
- diag_embed.out meta

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Meta registrations, Nov 7 edition, part 1"


I already had to fix a bunch of bugs in this batch, so landing it for
now before I tackle the next batch.

Billing of changes:
- diag.out is now composite
- randint layout/pin_memory bugfix
- unsqueeze_ meta
- bernoulli metas
- mul_/div_/logical_{and_,or_,xor_} metas
- add_/sub_ metas with alpha support
- diag_embed.out meta

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ngimel commented on torch/_meta_registrations.py:
```diff
@@ -1461,7 +1518,7 @@ def maybe_wrap_dim(dim: int, dim_post_expr: int, wrap_scalar: bool = True):
     min = -dim_post_expr
     max = dim_post_expr - 1
-    assert not (dim < min or dim > max)
+    assert not (dim < min or dim > max), f"dim {dim} out of bounds ({min}, {max})"
```
should we leave a TODO here to switch to canonicalize_dim from primTorch instead of maybe_wrap_dim? `canonicalize_dim` properly throws index error.

ezyang replied:
sure can do. One thing I will say is that it's nice having the names line up exactly with C++, makes porting easier

ngimel replied:
Should we then use `maybe_wrap_dim` instead of `canonicalize` in primtorch?

ezyang replied:
The c++ api isn't designed intentionally, it just exists. I would be happy to port over to an intentional api if there is one lol. But I am also trying to minimize work lol

ngimel replied:
Hence a TODO, to minimize current work :-)

ezyang committed:
Update on "Meta registrations, Nov 7 edition, part 1"


I already had to fix a bunch of bugs in this batch, so landing it for
now before I tackle the next batch.

Billing of changes:
- diag.out is now composite
- randint layout/pin_memory bugfix
- unsqueeze_ meta
- bernoulli metas
- mul_/div_/logical_{and_,or_,xor_} metas
- add_/sub_ metas with alpha support
- diag_embed.out meta

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Meta registrations, Nov 7 edition, part 1"


I already had to fix a bunch of bugs in this batch, so landing it for
now before I tackle the next batch.

Billing of changes:
- diag.out is now composite
- randint layout/pin_memory bugfix
- unsqueeze_ meta
- bernoulli metas
- mul_/div_/logical_{and_,or_,xor_} metas
- add_/sub_ metas with alpha support
- diag_embed.out meta

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Meta registrations, Nov 7 edition, part 1"


I already had to fix a bunch of bugs in this batch, so landing it for
now before I tackle the next batch.

Billing of changes:
- diag.out is now composite
- randint layout/pin_memory bugfix
- unsqueeze_ meta
- bernoulli metas
- mul_/div_/logical_{and_,or_,xor_} metas
- add_/sub_ metas with alpha support
- diag_embed.out meta

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang commented:
This has been split into small commits as I cannot diagnose the build failures on the big commit
